### ISSUE TITLE
Support and/or make clear how to make bindings via reflection

### ISSUE THREAD

**Issue by kinghajj** [Kodein-Framework/Kodein-DI#78]
We have a REST API implemented with DropWizard. It uses the resource/repository pattern. We currently aren't using any DI, which is starting to get rather hairy as we keep adding more. I've been experimenting using Kodein, and had hoped that I could use reflection to dynamically find all classes that inherit from our `Repository`/`Resource` interfaces and set bindings dynamically.

            val kodein = Kodein {
                val reflections = Reflections("com.stuff.api")

                bind() from setBinding<Resource>()

                val resourceClasses = reflections.getSubTypesOf(Resource::class.java).filter { !Modifier.isAbstract(it.modifiers) }
                for (resourceClass in resourceClasses) {
                    // doesn't work
                    bind() from singleton { resourceClass.getConstructor(Kodein::class.java).newInstance(kodein) }
                    bind<Resource>().inSet() with provider { kodein.Instance(TT(resourceClass)) }
                }
            }

The problem with the line under the comment is that the binding is made to `Resource` instead of the specific `T` that `resourceClass` represents, so the second iteration of the loop causes an exception. I found `Bind(TypeToken)` and started trying something like

    for (resourceClass in resourceClasses) {
        val typeToken = TT(resourceClass)
        Bind(typeToken). // what now?
    }

The problem, though, is that all of the implementations of the `Binding` interface take a type parameter, which I can't provide because of using reflection.

Maybe I'm missing something, and there's a better way to go about this? I'd really just like to avoid needing to write this for each of our dozens of resources:

    bind() from singleton { FooResource(kodein) }
    bind<Resource>().inSet() with provider { kodein.instance<FooResource>() }

**Comment by SalomonBrys** [comment]
As you found out, all the methods that start with a lowercase letter such as `bind` and `singleton` are inline methods that use type inference to infer the type being bound.

Now, there is a bug in version 4.1 that prevents to do "correctly" what you want (I forgot a couple of `out`s).

When 4.2 will be out, you will be able to do that:

```kotlin
Bind(TT(resourceClass)) with SingletonBinding<Resource>(TT(resourceClass)) { resourceClass.getConstructor(Kodein::class.java).newInstance(kodein) }
```

In 4.1, the problem is that TT(resourceClass) is infered as `Class<out Resource>`, and the out is breaking things.
So while waiting for 4.2, the workaround is to remove it:

```kotlin
@Suppress("UNCHECKED_CAST")
val resourceTT = TT(resourceClass) as TypeToken<Resource>
Bind(resourceTT) with SingletonBinding<Resource>(resourceTT) { resourceClass.getConstructor(Kodein::class.java).newInstance(kodein) }
```

Sorry about that, thank you for reporting it ;)

**Comment by SalomonBrys** [comment]
This is corrected in [Kodein 5.0-Beta1](https://salomonbrys.github.io/Kodein/?5.0).
I will close this issue once the stable 5.0 will be released.

**Comment by SalomonBrys** [comment]
Kodein 5.0 stable has been released :)
